core/src/consensus.rs

```diff
@@ -418,8 +520,8 @@ impl Tower {
                 }
 
                 // Should never consider switching to an ancestor
-                // of your last vote
-                assert!(!last_vote_ancestors.contains(&switch_slot));
+                // of your last vote unless the ancestor is restored from stray slots
+                assert!(is_restored || !last_vote_ancestors.contains(&switch_slot));
```

Comment:
This is needed for test_no_optimistic_confirmation_violation_with_tower. I'll look into further why this is needed later...